mm/pmm.c

```diff
@@ -150,16 +150,26 @@ void init_pmm(void) {
         display_frames();
 }
 
-static frame_t *reserve_frame(frame_t *frame, unsigned int order) {
-    BUG_ON(!frame);
-    BUG_ON(!frame->flags.free);
-    frame->flags.free = false;
+static inline frame_t *reserve_frame(frame_t *frame) {
+    if (!frame)
+        return NULL;
 
-    BUG_ON(frame->refcount > 0);
-    frame->refcount++;
+    if (frame->refcount++ == 0) {
+        list_unlink(&frame->list);
+        list_add(&frame->list, &busy_frames[frame->order]);
+    }
 
-    list_unlink(&frame->list);
-    list_add(&frame->list, &busy_frames[order]);
+    return frame;
+}
+
+static inline frame_t *return_frame(frame_t *frame) {
+    if (!is_frame_used(frame))
+        panic("PMM: trying to return unused frame: %p\n", frame);
+
+    if (--frame->refcount == 0) {
+        list_unlink(&frame->list);
+        list_add(&frame->list, &free_frames[frame->order]);
+    }
 
     return frame;
```

Comment:
Without understanding the details, Are we supposed to return `frame` even when refcount reached to zero?

Reply:
And why're we event returning it if it's the same frame we're returning?

Reply:
> Without understanding the details, Are we supposed to return `frame` even when refcount reached to zero?

We should return frame (i.e. re-assign to free_frames list) only when refcount is down to 0.

Reply:
> And why're we even returning it if it's the same frame we're returning?

I guess you mean, why the same frame pointer is returned from this function?
It is just handy if we want to do stuff like: `return return_frame(frame)`

Reply:
makes sense